source/merge-deep.d.ts

```diff
@@ -0,0 +1,396 @@
+				: // Assume recordMergeMode = 'replace'
+					Source[Key]
+			: // Only the source exists
+				Source[Key]
+		: // No source, take the destination
+		Key extends keyof Destination
+		? Destination[Key]
+		: never; // (this test is useless, but make TS happy, It can never be never)
+};
+
+/**
+Wrapper around {@link DoMergeRecord} which defines whether or not to strip `undefined` values.
+*/
+type MergeDeepRecord<
+	Destination,
+	Source,
+	Options extends MergeDeepOptions,
+> = Options['stripUndefinedValues'] extends true
+	? ConditionalExcept<DoMergeRecord<Destination, Source, Options>, undefined>
+	: DoMergeRecord<Destination, Source, Options>;
+
+/**
+Try to merge two objects or two arrays/tuples recursively into a new type or return the default value.
+*/
+type MergeDeepOrReturn<DefaultValue, Destination, Source, Options> = Destination extends UnknownArrayOrTuple
+	? Source extends UnknownArrayOrTuple
+		? ConditionalSimplify<MergeDeepArray<Destination, Source, Merge<MergeDeepDefaultOptions, Options>>, Function>
+		: DefaultValue
+	: Destination extends UnknownRecord
+	? Source extends UnknownRecord
+		? ConditionalSimplify<MergeDeepRecord<Destination, Source, Merge<MergeDeepDefaultOptions, Options>>, Function>
+		: DefaultValue
+	: DefaultValue;
+
+/**
+Merge mode for object properties.
+
+@see MergeDeepOptions.recordMergeMode
+*/
+export type MergeDeepRecordMode = 'merge-or-replace' | 'merge-or-union' | 'replace' | 'union';
+
+/**
+Merge mode for array/tuple items.
+
+@see MergeDeepOptions.arrayMergeMode
+*/
+export type MergeDeepArrayMode = MergeDeepRecordMode | 'merge-or-spread' | 'spread' | 'merge';
+
+/**
+MergeDeep options.
+
+@see MergeDeep
+*/
+export interface MergeDeepOptions {
+	/**
+	Should strip `undefined` values from the resulting type.
+
+	@default false
+	*/
+	stripUndefinedValues?: boolean;
+
+	/**
+	Merge mode for object properties.
+
+	See {@link MergeDeep} for usages and examples.
+
+	@default 'merge-or-replace'
+	*/
+	recordMergeMode?: MergeDeepRecordMode;
+
+	/**
+	Merge mode for array/tuple items.
+
+	See {@link MergeDeep} for usages and examples.
+
+	@default 'spread'
+	*/
+	arrayMergeMode?: MergeDeepArrayMode;
+}
+
+// Default options that will be merged with user provided options.
+interface MergeDeepDefaultOptions extends MergeDeepOptions {
+	stripUndefinedValues: false;
+	recordMergeMode: 'merge-or-replace';
+	arrayMergeMode: 'spread';
+}
+
+/**
+Merge two objects or two arrays/tuples recursively into a new type.
+
+@see MergeDeepOptions
+
+@experimental This type is marked as experimental because it depends on {@link ConditionalSimplify} which itself is experimental.
+
+@example
+```
+import type {MergeDeep} from 'type-fest';
+
+type Foo = {
+	life: number;
+	items: string[];
+	a: {b: string; c: boolean};
+};
+
+interface Bar {
+	name: string;
+	items: number[];
+	a: {b: number};
+}
+
+type FooBar = MergeDeep<Foo, Bar>;
+// FooBar = {
+// 	life: number;
+// 	name: string;
+// 	items: (string | number)[];
```

Comment:
It's not clear to me why I would want it to be typed as either merged or replaced? Why wouldn't I choose one?

Reply:
As I started from a version that only did the fusion on one level and added the rest after that branch exists and so I exposed it. In fact it has the same effect as `Merge` plus the options of filtering `undefined` values and merging arrays on a single level.